### ISSUE TITLE
Access is denied during removal of snapshots in containerd

### ISSUE THREAD

**Issue by jsturtevant** [microsoft/Windows-Containers#358]
**Describe the bug**

In kubernetes during image Garbage Collection, we are seeing the following logs in containerd: 

```
containerd.err.log:time="2023-04-12T09:31:05.002389400Z" level=warning msg="snapshot garbage collection failed" error="second rename attempt following detach failed: rename C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\1352 C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\rm-1352: Access is denied.: rename C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\1352 C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\rm-1352: Access is denied." snapshotter=windows
```

It looks to be happening during the OS removing layers leaving handles behind: https://github.com/kubernetes/kubernetes/issues/116020#issuecomment-1504991936

**To Reproduce** 
lots of details in https://github.com/kubernetes/kubernetes/issues/116020

**Expected behavior**
Image files should be deleted when the call to remove this is processed.

**Configuration:**
 - Edition: [e.g. Windows 11, Windows Server] windows server
- Base Image being used: [e.g. Windows Server Core]
 - Container engine: [e.g. docker] containerd
 - Container Engine version [e.g. 22]

**Additional context**

**Comment by sufuf3** [comment]
Hi, 
Thank you for working on the issue. 
This may need to take time for fix, for short-term do you have any workaround? Otherwise, this may cause some impacts from user side. 
Many thanks

**Comment by kiashok** [comment]
Hi! We are looking into this issue and hopefully should have an update soon. This also seems to be happening only on process isolated containers so an immediate work around might be to use hyperV isolated containers temporarily while this is fixed.
@sufuf3 is this currently a blocker for you guys?

**Comment by kiashok** [comment]
> Hi, Thank you for working on the issue. This may need to take time for fix, for short-term do you have any workaround? Otherwise, this may cause some impacts from user side. Many thanks

Hi @sufuf3 , do you have an active repro of this? Would you be able to share the repro steps?

**Comment by lippertmarkus** [comment]
@sufuf3 we're just monitoring the windows disks and replacing the nodes as soon as they get full.

@kiashok a repro from us you might already got from @gabriel-samfira together with the initial report :)

**Comment by kiashok** [comment]
> @sufuf3 we're just monitoring the windows disks and replacing the nodes as soon as they get full.
> 
> @kiashok a repro from us you might already got from @gabriel-samfira together with the initial report :)

@lippertmarkus the credentials to access the image have now expired. Could you reshare them through @gabriel-samfira?

**Comment by AbelHu** [comment]
@kiashok Maybe this fix https://github.com/containerd/containerd/pull/8414 may also address this issue. One AKS customer does not see full disk issue after upgrading their clusters (Windows node image: AKSWindows-2022-containerd-20348.1787.230614, Windows containerd: v1.6.21)

**Comment by lippertmarkus** [comment]
> @lippertmarkus the credentials to access the image have now expired. Could you reshare them through @gabriel-samfira?

@kiashok I reshared them through him, thanks!

**Comment by shiva-appani-hash** [comment]
> @kiashok Maybe this fix [containerd/containerd#8414](https://github.com/containerd/containerd/pull/8414) may also address this issue. One AKS customer does not see full disk issue after upgrading their clusters (Windows node image: AKSWindows-2022-containerd-20348.1787.230614, Windows containerd: v1.6.21)

Node image upgrade doesn't solve the problem, disk space issue persists.

OS
windows (amd64)
OS Image
Windows Server 2022 Datacenter
Kernel version
10.0.20348.1787
Container runtime
containerd://1.6.21+azure
Kubelet version
v1.24.9
node-image-version
AKSWindows-2022-containerd-20348.1787.230614

@AbelHu Please help.

**Comment by kiashok** [comment]
The root cause for this bug has been identified and we have a fix for the same. It should go out soon.

**Comment by lippertmarkus** [comment]
@kiashok thanks! Could you please give more infos on the version that will include the fix as soon as you have them? KB/build version and maybe AKS image version?

**Comment by jsturtevant** [comment]
As soon as we have dates/versions will share them here

**Comment by AbelHu** [comment]
I got the info that the fix will be included in the monthly Windows security patch in Sep, 2023.

**Comment by shiva-appani-hash** [comment]
@AbelHu Requesting you to share the release calendar or exact dates for East US, N.Europe and S.E.Asia deployment.

**Comment by TBBle** [comment]
@shiva-appani-gep: Per [Windows quality updates primer](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/windows-quality-updates-primer/ba-p/2569385) and [Windows Server 2022 update history](https://support.microsoft.com/en-gb/topic/windows-server-2022-update-history-e1caa597-00c5-4ab9-9f3e-8212fe80b2ee) it should be September 12th, [10am in Redmond](https://learn.microsoft.com/en-us/windows/deployment/update/release-cycle#monthly-security-update-release) (although this link is for clients). Here's a [local time conversion](https://www.timeanddate.com/worldclock/fixedtime.html?msg=Patch+Tuesday+September+2023&iso=20230912T10&p1=1244).

**Comment by AbelHu** [comment]
@shiva-appani-gep AKS needs some time to build, test and publish new AKS Windows images with 2023.09B but it should not more than 1 week after 9/12.

**Comment by TBBle** [comment]
Did the fix land in 20348.1970? I don't see anything I recognise as relevant in [the KB article](https://support.microsoft.com/en-gb/topic/september-12-2023-kb5030216-os-build-20348-1970-34d4aff3-fd05-4270-b288-4ab6379c7f81).

**Comment by shiva-appani-hash** [comment]
> @shiva-appani-gep AKS needs some time to build, test and publish new AKS Windows images with 2023.09B but it should not more than 1 week after 9/12.

@AbelHu I have gone through the KB article(https://support.microsoft.com/en-gb/topic/september-12-2023-kb5030216-os-build-20348-1970-34d4aff3-fd05-4270-b288-4ab6379c7f81) and couldn't any relevant information regarding this fix. Did we get the fix in 20348.1970, please confirm.

**Comment by AbelHu** [comment]
@shiva-appani-gep The fix is included in AKS Windows images with 2023.09B. Please file an ICM if you still see it with AKS Windows images with 2023.09B.

**Comment by kiashok** [comment]
This issue has now been fixed in 10B release. Closing this issue.
